pypeit/reduce.py

```diff
@@ -11,116 +11,96 @@
-           Set to 'science'
-         sciframe : ndarray
-           Processed 2D frame
-         rawvarframe : ndarray
-           Variance generated without a sky (or object) model
-         modelvarframe : ndarray
-           Variance generated with a sky model
-         finalvar : ndarray
-           Final variance frame
-         global_sky : ndarray
-           Sky model across the slit/order
-         skycorr_box : ndarray
-           Local corrections to the sky model
-         final_sky : ndarray
-           Final sky model; may include 'local' corrections
-         obj_model : ndarray
-           Model of the object flux
-         trcmask : ndarray
-           Masks of objects for sky subtraction
-         tracelist : list
-           List of traces for objects in slits
-         inst_name : str
-           Short name of the spectrograph, e.g. KASTb
-         target_name : str
-           Parsed from the Header
-         basename : str
-           Combination of camera, target, and time
-           e.g. J1217p3905_KASTb_2015May20T045733.56
-         time : Time
-           time object
-         specobjs : list
-           List of specobjs
-         bm: ScienceImageBitMask
-           Object used to select bits of a given type
-     """
+    This class will organize and run actions related to
+    finding objects, sky subtraction, and extraction for
+    a Science or Standard star exposure
+
+    Args:
+        sciImg (pypeit.images.scienceimage.ScienceImage):
+        spectrograph (pypeit.spectrograph.Spectrograph):
+        par (pypeit.par.pyepeitpar.PypeItPar):
+        caliBrate (pypeit.calibrations.Calibrations):
+           This is only used as a container and it must contain the main products
+           of WaveTilts, WaveImage, and EdgeTrace
+        det (int, optional):
+           Detector indice
+        setup (str, optional):
+           Used for naming
+        maskslits (ndarray, optional):
+          Specifies masked out slits
+          True = Masked
+        objtype (str, optional):
+           Specifies object being reduced 'science' 'standard' 'science_coadd2d'
+        show (bool, optional):
+           Show plots along the way?
+
+    Attributes:
+        ivarmodel (np.ndarray):
+            Model of inverse variance
+        objimage (np.ndarray):
+            Model of object
+        skyimage (np.ndarray):
+            Final model of sky
+        initial_sky (np.ndarray):
+            Initial sky model after first pass with global_skysub()
+        global_sky (np.ndarray):
+            Fit to global sky
+        skymask (np.ndarray):
+            Mask of the sky fit
+        outmask (np.ndarray):
+            Final output mask
+        extractmask (np.ndarray):
+            Extraction mask
+        sobjs_obj (SpecObjs):
+            Only object finding but no extraction
+        sobjs (SpecObsj):
+            Final extracted object list with trace corrections applied
+
+    """
 
     __metaclass__ = ABCMeta
 
```

Comment:
I think par should be optional, since you can pull the defaults out of spectrograph if it is not passed. Can you add something to the docs explaining what attributes in calibrate need to be set, i.e. basically just the ugly incantation you had to add to coadd2d to create the calibrate object.

Reply:
By the way at some level, tslits_dict, tilts, and waveimage could live in a different object than flats, biases, and bpms. The latter are all used to create a sciImg object, whereas the former are required for a reduction. That would remove the ambiguity of what needs to be set in caliBrate to run a reduction. 

The reason I'm being anal about this is I'd like these classes to be usable in a Notebook with a few lines of code say if you want to check some part of a reduction.

Reply:
I do think we need a different approach, but we'll need to think
through what that is.  And better to do so once we have our
data models fixed and in place.  As for the comment about `par`,
that is too big a change in philosophy to institute here.